### PR TITLE
Finite State Machine

### LINKED ISSUE #15: Finite State Machine
Is your feature request related to a problem? Please describe.
Lack of a easy way to implement Finite State Machines.

Describe the solution you would like
Base classes which can be used to build a Finite State Machine.

### PR DESCRIPTION
**Pull request by gnoriega27**
## What issue is this PR solving?
Closes: #15 

## Description of change
Simple finite state machine with an example.

## How to reproduce the changes
1. Open the FSM scene.
2. Start.
3. Click anywhere in the plane.
4. While the player is moving you can press "spacebar" and it will run.
5. Observer the state changes in the textview.

## Screenshots or GIFs
If applicable, show off your work!

### CODE REVIEW

File: Assets/Scripts/FSM/State.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            set => _stateName = value;
+        }
+
+        protected virtual void Start()
+        {
+            this._stateMachine = GetComponent<StateMachine>();
+        }
+
+        /// <summary>
+        /// Callback when that is called when you enter the state.
```
Comment (mjimenez98) [inline Assets/Scripts/FSM/State.cs]:
```suggestion
        /// Callback that is called when you enter the state.
```

File: Assets/Scripts/FSM/State.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        /// <summary>
+        /// Main function of the state. Called on every Update(). Must be overridden.
+        /// </summary>
+        public virtual void Execute()
+        {
+        }
+
+        /// <summary>
+        /// Callback when that is called when you exit the state.
```
Comment (mjimenez98) [inline Assets/Scripts/FSM/State.cs]:
```suggestion
        /// Callback that is called when you exit the state.
```

File: Assets/Scripts/FSM/State.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using UnityEngine;
+
+namespace Com.StudioTBD.CoronaIO.FMS
+{
+    public abstract class State : MonoBehaviour
+    {
+        /// <summary>
+        /// State machine where that this state belongs to.
```
Comment (mjimenez98) [inline Assets/Scripts/FSM/State.cs]:
```suggestion
        /// State machine to which this state belongs.
```

**Review by mjimenez98 — APPROVED** [review]
I am loving this, great work!

I made a few petty suggestions for clearer comments but other than that, this couldn't be better.

If you feel like writing about this architecture and how to use it in the README, please feel welcome to.

giphy

**Review by mjimenez98 — CHANGES_REQUESTED** [review]
I just noticed the .idea/ files. I can remove these. Putting the approval on hold until then.

**Review by mjimenez98 — APPROVED** [review]
We are good to go!

**Review by ghost — APPROVED** [review]
This sucks. Merge it.

**Comment by gnoriega27** [comment]
Todo:
- [ ] Add transition abstraction
- [ ] Add event listener abstraction

**Comment by mjimenez98** [comment]
Going over it again, I was trying to find problems with the code but couldn't find any. So that's great! 

My only question is, if the behavior in the scene _FSM_ is already being demonstrated in _HumanAgent_, could we delete the former so we have less things in the project?

**Comment by gnoriega27** [comment]
@mjimenez98 They are showing different examples. One is purely a FMS, while the other is the abstraction of an agent that has a FSM.

**Pull request merged**